### ISSUE TITLE
'No such column' when updating entity that has many-to-many relation

### ISSUE THREAD

**Issue by voducchung** [typeorm/typeorm#1987]
**Issue type:**

[ ] question
[x] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

[ ] `cordova`
[ ] `mongodb`
[ ] `mssql`
[ ] `mysql` / `mariadb`
[ ] `oracle`
[ ] `postgres`
[x] `sqlite`
[ ] `sqljs`
[ ] `react-native`

**TypeORM version:**

[ ] `latest`
[ ] `@next`
[x] `0.1.21`

**Steps to reproduce or a small repository showing the problem:**
I have 3 tables: 

`Product` (`id`: primary key), `Tag` (`id`: primary key) and `Tagging` (`id`: primary key, `catalog_id`: refers to product id, `tag_id`: refers to tag id).
```typescript
@Entity()
export class Product {
    @PrimaryColumn('uuid')
    @Generated('uuid')
    id: string

    @ManyToMany(type => Tag)
    @JoinTable({
        name: 'tagging',
        joinColumn: { name: 'catalog_id', referencedColumnName: 'id' },
        inverseJoinColumn: { name: 'tag_id', referencedColumnName: 'id' }
    })
    tags: Tag[]
}
```
```typescript
@Entity()
export class Tag {
    @PrimaryColumn('uuid')
    @Generated('uuid')
    id: string
}
```
```typescript
@Entity()
export class Tagging {
    @PrimaryColumn('uuid')
    @Generated('uuid')
    id: string

    @Column('text', { name: 'catalog_id' })
    catalogId: string

    @Column('text', { name: 'tag_id' })
    tagId: string
}
```

When I update a product with this code,
```typescript
let product = new Product()
product.id = productId
product.barCode = barCode
product.name = name

await getConnection('default').getRepository(Product).save(product)
```
I get this error:
```
{ QueryFailedError: SQLITE_ERROR: no such column: subject_tags_relation_id.tag_id.id
    at new QueryFailedError (E:\Personal\Experimenal\my-store-middleware\src\error\QueryFailedError.ts:7:9)
    at E:\Personal\Experimenal\my-store-middleware\src\driver\sqlite\SqliteQueryRunner.ts:54:26
    at replacement (E:\Personal\Experimenal\my-store-middleware\node_modules\sqlite3\lib\trace.js:19:31)
    at Statement.errBack (E:\Personal\Experimenal\my-store-middleware\node_modules\sqlite3\lib\sqlite3.js:16:21)
  message: 'SQLITE_ERROR: no such column: subject_tags_relation_id.tag_id.id',
  errno: 1,
  code: 'SQLITE_ERROR',
  name: 'QueryFailedError',
  query: 'SELECT subject_tags_relation_id.tag_id.id AS "tag_id", subject_tags_relation_id.catalog_id.id AS "catalog_id" FROM "tag" "tag" INNER JOIN "tagging" "subject_tags_relation_id" ON (subject_tags_relation_id.catalog_id.id = $1 AND subject_tags_relation_id.tag_id.id = "tag"."id") ORDER BY subject_tags_relation_id.tag_id.id ASC, subject_tags_relation_id.catalog_id.id ASC',
  parameters: [ '73d8fc9c-c252-48ae-8b0b-f0d60134a853' ] }
Debug: internal, implementation, error
    Error: handler method did not return a value, a promise, or throw an error
    at module.exports.internals.Manager.execute (E:\Personal\Experimenal\my-store-middleware\node_modules\hapi\lib\toolkit.js:52:29)
    at <anonymous>
```

The **special note** is that, if I remove the line ```product.id = productId``` (meaning insert instead of update), there is no error thrown

**Comment by voducchung** [comment]
This issue doesn't occur with typeorm 0.2.0. Closing this issue.

**Comment by GitBalou** [comment]
I think this issue needs to be reopen, i have the same thing with 0.2.6.
``````

**Comment by pleerock** [comment]
@GitBalou please create a separate PR with a failing test attached

**Comment by darkyelox** [comment]
@pleerock you must stop asking for PR with failing test in every issue because some people don't have the time for doing it (they already use their time reporting and pasting code here), i saw already too many unresolved (and worst closed issues) because you always ask for PR and that is the last response, you have everything in this issue for recreate the problem in somewhere with latest version and see what is happening.

I have this problem in version 0.2.11 too

**Comment by vlapo** [comment]
Hi @darkyelox. 

This is open source library and we and specially pleerock do everything in free time. Is it so bad to ask for some lead in form of PR with failing test? I think it is not.  Everybody want to use but just someone want to contribute and help. So start contribute and you will see how hard is help everyone.

**Comment by pleerock** [comment]
@darkyelox if you have this issue then please create a PR with a failing test. And also if issue exists please try to fix it. Also it would be really grateful if you create PR with a failing test and fix if possible for any exist 500+ opened issues. Everybody including people like you will appreciate it a lot. This is how it works, because it's free.

**Comment by darkyelox** [comment]
@vlapo i know what is an open source library but the same thing i did said, no everyone that post some issue here have the time for create PR's and, as you said, is hard to help everyone, i'm an user for this wonderful library but i work in my own projects and in this time i don't have to time for seeing others code and try to guessing what is wrong, at the moment i'm the kind of user that report issues and paste here the code involving it but no more (for someone with time and knowledge that can test the issue in their already cloned project and find a solution and creates the PR).

@pleerock as my last comment i repeat: not everyone have the time for creating a PR when publishing and issue and worst try to find a solution (the time for clone this project, read the code, try to guess what it do, try to guess what is wrong and finally try find a solution is too much time involved) and you as member i guess known the code better.

What i try to say is that in this library i see some issues (that ones that occurs to me) that are closed and the people claim to reopen but is like no one cares and still closed.

I'm only giving an advice about what i thing is wrong with the method for solve some issues in this repository

**Comment by Kononnable** [comment]
@darkyelox 
I agree with you partially. 
As a user I find a bug, search if it's described, if not make an issue. This is pretty similar in every case(OSS or non-OSS).
Now we have several paths how to proceed:
1) do nothing - User is happy because he feels like he helped. Some even from time to drops by just to say that it should be fixed or to add `+1`.
2) only respond to questions in PR(on at least weekly basis) - At this point issue would be really taken care of in a world which is close to perfect.
3) make a PR with failing test - Time spent on fixing bug by person who knows project internals is halved.
4) make a fix - Takes most time from user, greatest effect, nothing really to discuss here.

Why 1) is mostly worthless? Most of the time it doesn't contain necessary data to reproduce the issue, waiting for answers to questions how to reproduce take a lot of time. Person willing to fix the issue will forget 2 times what the issue was about before starting to actually fix it. 

Why such issues(1,2) are stalled and often closed if opportunity rise? Because a state of the software. Truth be told typeorm(in this case) is far from being ideal. It provides functionality but have a lot of bugs. So we have to prioritize which to fix(we don't have enough time to fix all of them) and which  to ignore for the time being. And here comes calculations. Would you like to fix two bugs of (3) or one bug of (1,2)? You have limited resources(mostly time), so you can't do both.

Why such issues(1,2) are often closed? Maintenance costs. If you have 100 issues you can pretty much remember all of them. Not details, but you look at title and just know what it is about. With 500 it is not even a possibility. Without closing such(1,2) issues we would have over 1k open issues right now. Owners wouldn't even care about half of them and we would waste a lot of time on maintenance or  fixing issues which nobody even cares about anymore. 

**To sum up - issues purpose isn't to make typeorm ideal. It is to make working with typeorm better.**
Since we don't live in an utopia we have to let go some of issues(or maintenance of them would be hell). Are we happy about that? No. Are users happy about that? No. But I think if we know we won't fix some rarely seen issue in near future then such issue should be closed. If it is crucial for someone he will write a test - issue will go to (3) and will have much greater chance of being fixed. If not, then issue wasn't so important. Otherwise issue owner would wait half a yer just thinking that someone will finally fix the issue.

Do we give great experience to the user? No, but  I have no idea how to do it better with current resources.

PS: I put 1, 2 in one group most of the time. It's because person which can fix a bug have to decide is it some common bug, or it's very specific to user code. And that's why some issues with almost no description stay alive and other doesn't.
